tests_e2e/orchestrator/lib/agent_test_suite_combinator.py

```diff
@@ -118,160 +119,215 @@ def _next(self) -> Optional[Dict[str, Any]]:
     def create_environment_for_existing_vm(self) -> List[Dict[str, Any]]:
         loader = AgentTestLoader(self.runbook.test_suites, self.runbook.cloud)
 
-        environment: List[Dict[str, Any]] = [
-            {
-                "c_env_name": self.runbook.vm_name,
-                "c_vm_name": self.runbook.vm_name,
-                "c_location": self.runbook.location,
-                "c_test_suites": loader.test_suites,
-            }
-        ]
+        environment: Dict[str, Any] = {
+            "c_env_name": self.runbook.vm_name,
+            "c_vm_name": self.runbook.vm_name,
+            "c_location": self.runbook.location,
+            "c_test_suites": loader.test_suites,
+        }
 
         log: logging.Logger = logging.getLogger("lisa")
-        log.info("******** Environment for existing VMs *****")
-        log.info(
-            "{  c_env_name: '%s', c_vm_name: '%s', c_location: '%s', c_test_suites: '%s' }",
-            environment[0]['c_env_name'], environment[0]['c_vm_name'], environment[0]['c_location'], [s.name for s in environment[0]['c_test_suites']])
-        log.info("***************************")
+        log.info("******** Waagent: Settings for existing VM *****")
+        log.info("")
+        log.info("Settings for %s:\n%s\n", environment['c_env_name'], self._get_env_settings(environment))
+        log.info("")
 
-        return environment
+        return [environment]
 
     def create_environment_list(self) -> List[Dict[str, Any]]:
+        """
+        Examines the test_suites specified in the runbook and returns a list of the environments (i.e. test VMs) that need to be
+        created in order to execute these suites.
+
+        Note that if the runbook provides an 'image', 'location', or 'vm_size', those values override any values provided in the
+        configuration of the test suites.
+        """
+        environments: List[Dict[str, Any]] = []
+        shared_environments: Dict[str, Dict[str, Any]] = {}  # environments shared by multiple test suites
+
         loader = AgentTestLoader(self.runbook.test_suites, self.runbook.cloud)
 
-        #
```

Comment:
I needed to touch this code to add the random selection, and this method was already way too long, so I refactored this logic into smaller functions.

The new logic is part of _get_test_suite_images().